contribute/design/1.x/cstor/20200511-cstor-backupandrestore.md

```diff
@@ -0,0 +1,264 @@
+CMD:
+   1. If request is for full backup then command is `zfs send <snapshot_dataset_name> | nc -w 3 <ip_address> <port>`
+   2. If request is for incremental backup then command is `zfs send -i <old_snapshot_dataset_name> <new_snapshot_dataset_name> | nc -w 3 <ip_address> <port>`
+4. Updates the corresponding CStorCompletedBackups with last two completed backups.
+   For example, if schedule `b` has last two backups b-0 and b-1 (b-0 created first and after that b-1 was created) having snapshots
+   b-0 and b-1 respectively then CStorCompletedBackups for the schedule `b` will have following information :
+```go
+   CStorCompletedBackups.Spec.PrevSnapName =  b-1
+   CStorCompletedBackups.Spec.SnapName = b-0
+```
+NOTE: ip_address and port are the IP Address and port of snapshot sender/receiver module.
+
+##### Work flow in restore controller
+
+When REST API `/latest/restore` is executed it creates CStorRestore CR with `Pending` status. Restore controller which present in pool-manager will get event and perform the following operations:
+1. Update the CStorRestore status to `Init`(which conveys controller instantiate process).
+2. Update the status of CStorRestore resource as `InProgress`(Which will help to understand the restore process is started).
+3. Execute the below command and ZFS will receive the data from `sender/receiver module`(blocking call and this command execution will be retried for 50 seconds at interval of 5 seconds in case of errors).
+CMD: `nc -w 3 <ip_address> <port> | zfs recv -F <volume_datasetname>`
+
+NOTE: ip_address and port are the IP Address and port of snapshot sender/receiver module.
+
+## Design Details
+
+### CStorBackup Schema
+Following is the existing CStorBackup schema in go struct:
+```go
+// CStorBackup describes a cstor backup resource created as a custom resource
+type CStorBackup struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              CStorBackupSpec   `json:"spec"`
+	Status            CStorBackupStatus `json:"status"`
+}
+
+// CStorBackupSpec is the spec for a CStorBackup resource
+type CStorBackupSpec struct {
+	// BackupName is a name of the backup or scheduled backup
+	BackupName string `json:"backupName"`
+
+	// VolumeName is a name of the volume for which this backup is destined
+	VolumeName string `json:"volumeName"`
+
+	// SnapName is a name of the current backup snapshot
+	SnapName string `json:"snapName"`
+
+	// PrevSnapName is the last completed-backup's snapshot name
+	PrevSnapName string `json:"prevSnapName"`
+
+	// BackupDest is the remote address for backup transfer
+	BackupDest string `json:"backupDest"`
+
+	// LocalSnap is flag to enable local snapshot only
+	LocalSnap bool `json:"localSnap"`
+}
+
+// CStorBackupStatus is to hold status of backup
+type CStorBackupStatus string
```

Comment:
Let's add supported `cstorbackupstatus` and `CStorRestoreStatus` values.

Reply:
Added the various phases